[react-router] added possibility to specify component props in route definitions, and to specify the type of this.props.route.props when implementing such a component

neuoy opened the pull request:
Please note this is for react-router version 3, even though index.d.ts indicates "Type definitions for React Router 4.0" it's in the v3 folder, and seems to match with actual implementation of react-router 3

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
This feature does not seem to be documented, I found it here: https://github.com/mik-jozef/react-router9#custom-props (and though this is a different project, it does work in react-router too)
here is a code sample: https://github.com/neuoy/DefinitelyTyped/commit/2ebb416c8b09abc75cadc485bcae708ee0e37918
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

neuoy commented:
Sorry, didn't see the v3/react-router-tests.ts file, which fails, I'll take a look

andy-ms commented on types/react-router/v3/lib/IndexRoute.d.ts:
```diff
@@ -12,17 +12,18 @@ import {
 type ComponentsCallback = (err: any, components: RouteComponents) => any;
 
-export interface IndexRouteProps {
+export interface IndexRouteProps<Props> {
```
Try `<Props = any>`, otherwise this is a breaking change. Same for `RouteProps` and `PlainRoute`.

neuoy commented:
You're right, I adjusted the new generic type variables so that they default to `any`. I've also adjusted places where the modified types were used to remove the now useless `<any>` parameter. The whole thing is now simpler and more compatible.

Also I don't understand why the lint checker is so picky about Array<T> versus T[], since it seems to work fine in typescript (it does in my project anyway). My bad for not running lint locally before pushing though. This is fixed now.

I've done and undone a few things in those successive commits, I could make a new pull-request with less commits if you wish, let me know.

andy-ms commented:
Thanks!
